pyuvm/s18_pyuvm_reg_block.py

```diff
@@ -1,19 +1,182 @@
+        self.maps = []
+        self.parent_blk = None
+        # optimize this by using dicationries with
+        # TUPLES (existence_bool, handle)
+        self.blk_maping = {}
+        self.map_mapping = {}
+        self.reg_mapping = {}
+        self.blk_name = name
+        self.header = name + " -- "
 
-    # 18.1.3.7
-    # TODO Fix signature
-    def get_registers(self):
-        return self._regs
+    # is_locked
+    def is_locked(self) -> bool:
+        return self._is_locked
 
+    # gen_message
+    def gen_message(self, txt="") -> str:
+        return f"{self.header,txt}"
+
+    # clear_hdl_path
+    def clear_hdl_path(self, kind="RTL"):
+        uvm_not_implemeneted(self.header)
+
+    # add_hdl_path
+    def add_hdl_path(self, path: str, kind="RTL"):
+        uvm_not_implemeneted(self.header)
+
+    # has_hdl_path
+    def has_hdl_path(self, kind: str) -> bool:
+        uvm_not_implemeneted(self.header)
+
+    # get_hdl_path
+    def get_hdl_path(self, paths: list, kind=""):
+        uvm_not_implemeneted(self.header)
+
+    # get_full_hdl_path
+    def get_full_hdl_path(self, paths: list, kind="", separator="."):
+        uvm_not_implemeneted(self.header)
+
+    # set_lock
+    def set_lock(self):
+        self._is_locked = True
+
+    # blk_set_reg_mapping
+    def blk_set_reg_mapping(self, reg: uvm_reg):
+        self.reg_mapping[reg.get_name()] = True
+
+    # blk_is_reg_mapped
+    def blk_is_reg_mapped(self, reg: uvm_reg) -> bool:
+        return self.reg_mapping[reg.get_name()]
+
+    # blk_is_child_mapped
+    def blk_is_child_mapped(self, in_blk) -> bool:
+        if (isinstance(in_blk, uvm_reg_block) is False):
+            uvm_fatal(self.gen_message("blk_is_child_mapped -- input block should be uvm_reg_block"))
+        else:
+            return self.blk_maping[in_blk.get_name()]
+
+    # blk_set_map_mapping
+    def blk_set_map_mapping(self, map_i: uvm_reg_map):
+        self.map_mapping[map_i.get_name()] = 1
+
+    # blk_is_map_mapped
+    def blk_is_map_mapped(self, map_i: uvm_reg_map):
+        self.map_mapping[map_i.get_name()] = 1
+
+    # configure
+    def configure_blk(self, parent, hdl_path):
+        if (self.parent_blk is None):
+            self.parent_blk = parent
+            self.parent_blk.configure_blk(self)
+        else:
+            self.parent_blk = None
+        # add HDL PATH as well
+        self.add_hdl_path(hdl_path)
+
+    # add_block
+    def add_block(self, in_blk):
+        if (isinstance(in_blk, uvm_reg_block) is False):
+            uvm_fatal(self.gen_message("add_block -- input block must be uvm_reg_block type"))
+        # add to the BLK main mapping
+        if (in_blk not in self.child_blk):
+            self.blk_maping[in_blk.get_name()] = 1
+            self.child_blk.append(in_blk)
+
+    # blk_add_register
     def _add_register(self, reg):
-        self._regs.append(reg)
\ No newline at end of file
+        self._regs.append(reg)
+        self.blk_set_reg_mapping(reg)
```

Comment:
anytime a new register will be added we need to map it using the name so 2 registers with same name will not create aliasing